From 1059e8e0dadbc9f556ad89ae5208aa116e6aea06 Mon Sep 17 00:00:00 2001
From: Rob Clark <rclark@redhat.com>
Date: Wed, 17 Dec 2014 21:39:29 -0500
Subject: [drm] i915: Ignore long hpds on eDP ports

Message-id: <1418852380-13061-79-git-send-email-rclark@redhat.com>
Patchwork-id: 102216
O-Subject: [RHEL7 drm 78/89] drm/i915: Ignore long hpds on eDP ports
Bugzilla: 1173317
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Dave Airlie <airlied@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1173317
Upstream: since drm-intel-fixes-2014-10-30

commit 7a7f84ccb82e542c845c43f604665ccea1247866

Author:     Ville Syrjala <ville.syrjala@linux.intel.com>
AuthorDate: Thu Oct 16 20:46:10 2014 +0300
Commit:     Rob Clark <rclark@redhat.com>
CommitDate: Mon Dec 15 14:33:31 2014 -0500

    drm/i915: Ignore long hpds on eDP ports

    Turning vdd on/off can generate a long hpd pulse on eDP ports. In order
    to handle hpd we would need to turn on vdd to perform aux transfers.
    This would lead to an endless cycle of
    "vdd off -> long hpd -> vdd on -> detect -> vdd off -> ..."

    So ignore long hpd pulses on eDP ports. eDP panels should be physically
    tied to the machine anyway so they should not actually disappear and
    thus don't need long hpd handling. Short hpds are still needed for link
    re-train and whatnot so we can't just turn off the hpd interrupt
    entirely for eDP ports. Perhaps we could turn it off whenever the panel
    is disabled, but just ignoring the long hpd seems sufficient.

    Signed-off-by: Ville Syrjala <ville.syrjala@linux.intel.com>
    Cc: stable@vger.kernel.org
    Reviewed-by: Dave Airlie <airlied@redhat.com>
    Reviewed-by: Todd Previte <tprevite@gmail.com>
    Signed-off-by: Jani Nikula <jani.nikula@intel.com>

Signed-off-by: Rob Clark <rclark@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/i915/intel_dp.c b/drivers/gpu/drm/i915/intel_dp.c
index 61c370e..0cd7537 100644
--- a/drivers/gpu/drm/i915/intel_dp.c
+++ b/drivers/gpu/drm/i915/intel_dp.c
@@ -3929,6 +3929,18 @@ intel_dp_hpd_pulse(struct intel_digital_port *intel_dig_port, bool long_hpd)
  if (intel_dig_port->base.type != INTEL_OUTPUT_EDP)
   intel_dig_port->base.type = INTEL_OUTPUT_DISPLAYPORT;
 
+ if (long_hpd && intel_dig_port->base.type == INTEL_OUTPUT_EDP) {
+  /*
+   * vdd off can generate a long pulse on eDP which
+   * would require vdd on to handle it, and thus we
+   * would end up in an endless cycle of
+   * "vdd off -> long hpd -> vdd on -> detect -> vdd off -> ..."
+   */
+  DRM_DEBUG_KMS("ignoring long hpd on eDP port %c\n",
+         port_name(intel_dig_port->port));
+  return false;
+ }
+
  DRM_DEBUG_KMS("got hpd irq on port %d - %s\n", intel_dig_port->port,
         long_hpd ? "long" : "short");
 
-- 
1.7.1